File: src/main/java/com/coding/exam/Bfs.java
package com.coding.exam;

import java.util.*;

public class Bfs {
    public static void main(String[] args) {
        int[][] graph = {
                {1,2,4},
                {0, 2},
                {1,0,3},
                {2,4},
                {0,3,2}
        };
        List<List<Integer>> adjGraph = new ArrayList<>();
        adjGraph.add(0, Arrays.asList(1,2,4));
        adjGraph.add(1, Arrays.asList(0,2));
        adjGraph.add(2, Arrays.asList(1,0,3));
        adjGraph.add(3, Arrays.asList(2,4));
        adjGraph.add(4, Arrays.asList(0,3,2));



        boolean[] visited = new boolean[graph.length];

        bfs(graph, visited, 0);
        System.out.println();
        boolean[] visited1 = new boolean[adjGraph.size()];
        adjBfs(adjGraph, visited1, 0);
    }

    public static void bfs(int[][] graph, boolean[] visited, int start) {
        Queue<Integer> queue = new LinkedList<>();
        System.out.print(start+ " ");
        queue.add(start);
        visited[start] = true;
        while(!queue.isEmpty()) {
            int v = queue.poll();
            for (int vv : graph[v]) {
                if (!visited[vv]) {
                    System.out.print(vv + " ");
                    queue.add(vv);
                    visited[vv] = true;
                }
            }
        }

    }

    public static void adjBfs(List<List<Integer>> graph, boolean[] visited, int start) {
        Queue<Integer> queue = new LinkedList<>();
        System.out.print(start+ " ");
        queue.add(start);
        visited[start] = true;
        while(!queue.isEmpty()) {
            int v = queue.poll();
            for (int vv : graph.get(v)) {
                if (!visited[vv]) {
                    System.out.print(vv + " ");
                    queue.add(vv);
                    visited[vv] = true;
                }
            }
        }

    }
}

File: src/main/kotlin/com/coding/programmers/etc/BFS.kt
import java.util.*

class BFS {
    fun solution(graph: Array<IntArray>, start: Int, n: Int): Array<Int> {
        val adjList = Array<ArrayList<Int>>(n + 1) { ArrayList() }
        for (edge in graph) {
            adjList[edge[0]].add(edge[1])
        }

        return bfs(start, adjList)
    }

    fun bfs(start: Int, adjList: Array<ArrayList<Int>>): Array<Int> {
        val queue = ArrayDeque<Int>()
        val visited = mutableSetOf<Int>()
        val answer = mutableListOf<Int>()
        queue.offer(start)
        visited.add(start)

        while (queue.isNotEmpty()) {
            val vertex = queue.poll()

            answer.add(vertex)

            adjList[vertex].forEach {
                if (!visited.contains(it)) {
                    queue.offer(it)
                    visited.add(it)
                }
            }
        }

        return answer.toTypedArray()
    }
}